### ISSUE TITLE
Bug in Style/FrozenStringLiteralComment where comment can exist anywhere in file

### ISSUE THREAD

**Issue by jeffcarbs** [rubocop-hq/rubocop#7929]
Looking at this history of the `Style/FrozenStringLiteralComment` cop it looks like it:
- originally only checked the first three lines of source: https://github.com/rubocop-hq/rubocop/blob/58f2f8d37b733436bbc6cfa49594828699825de2/lib/rubocop/cop/mixin/frozen_string_literal.rb#L37
- was updated to check the first three comments but still looked at the LOC to see it was within the first three: https://github.com/rubocop-hq/rubocop/blob/75eb10738e356121bcf5dc6ce54d0ba45cd1e595/lib/rubocop/cop/mixin/frozen_string_literal.rb#L42-L50
- then was updated to simply check the first three comments (to handle cases where comments were separated by newlines): https://github.com/rubocop-hq/rubocop/commit/8c792d1349e7cdf383ebde08b00c83363d52db05

`frozen_string_literal` actually must appear before any code. Here's an example that should raise an exception at runtime but doesn't:
```ruby
1 + 2
# frozen_string_literal: true

K = "foo"
K.slice!(4)
```

However, since the cop is checking the first three comments they can appear anywhere in the source. For example, according to the cop these are OK:
```ruby
class Something
end

# frozen_string_literal: true
```

```ruby
# One

class Something
  # Two
end

# frozen_string_literal: true
```

This is actually invalid but "accidentally" because the cop has logic about this needing to be one of the first three comments:
```ruby
# One

class Something
  # Two
end

# Three
# frozen_string_literal: true
```

There's actually nothing special about the first three lines/comments either. This should be valid as well:
```ruby
# something else that's important
# I love comments
# comments commenty comments
# Oh boy, more comments

# frozen_string_literal: true
q = 1 + 2
K = "foo"
K.slice!(4)
```

The reason this cop cares about "first three" is because it has special logic dealing with the `shebang` and `encoding` comments.

--------

## Expected behavior

This cop would enforce the string literal comment is somewhere in the _leading_ comments (i.e. comments appearing before any ruby code), not necessarily just the first three.

## Actual behavior

The cop checks the first three comments in source, regardless of their location in the file

## Steps to reproduce the problem

See test cases above.

## RuboCop version

```
$ [bundle exec] rubocop -V
0.82.0 (using Parser 2.5.1.2, running on ruby 2.5.1 x86_64-linux)
```

**Comment by zhustec** [comment]
Consider a executable ruby file with Unix shebang

would it be 

```ruby
#!/usr/bin/env ruby
# frozen_string_literal: true
```

or

```ruby
# frozen_string_literal: true
#!/usr/bin/env ruby
```

**Comment by koic** [comment]
Shebang would always have to be on the first line.

```console
% cat example1.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
p 'str'.frozen?

% example1.rb
true

% cat example2.rb
# frozen_string_literal: true
#!/usr/bin/env ruby
p 'str'.frozen?

% example2.rb
example2.rb: line 3: p: command not found
```

See also: https://docs.rubocop.org/en/stable/cops_lint/#lintorderedmagiccomments

**Comment by jeffcarbs** [comment]
Fixed in #7939 🙂
